### DISCUSSION TITLE
Logged models in runs tab show initial not final metrics?

### DISCUSSION THREAD

**Discussion by Siniara** [mlflow/mlflow#20863] [Q&A]
I just noticed this and this confused me a lot. I train, log metrics per epoch, then save the model. Everything seems to go well, but then I noticed that my run metrics don't correspond at all with the metrics that mlflow associates with the particular model. For model saving I simply do: 
```python
mlflow.pytorch.log_model(
            model, name=model_type
        )
```
I was very confused, but then figured out that the metrics associated with the model are the metrics of the first epoch. I don't know why this happens, I'm sure that I log my model post training as when i load it it works/is trained. Is this expected, is this a bug?

From the particular run tab:
<img width="1005" height="594" alt="image" src="https://github.com/user-attachments/assets/04fe4f67-3eca-4718-97a1-128632b95aa2" />
Same experiment/tab, lower:
<img width="899" height="281" alt="image" src="https://github.com/user-attachments/assets/8446a22a-67ef-4113-974a-c0ac01762ef5" />

**Comment by omni-front** [comment]
ran into this myself, and it can definitely be a bit confusing. the behavior you're seeing is due to how metrics are logged in mlflow. when you log metrics in a loop, like per epoch during training, mlflow keeps track of all those values. but, when you view metrics under the "models" section, it typically shows the first logged value unless you specify otherwise.

one thing you can do is ensure you're logging the final metrics separately after training completes. here's a small tweak you can try:

```python
# log metrics during training
for epoch in range(num_epochs):
    # your training loop
    mlflow.log_metric("accuracy", epoch_accuracy, step=epoch)

# log the final metrics after training
mlflow.log_metric("final_accuracy", final_accuracy)
```

this way, when you look at the "models" section, you can have a specific metric like `final_accuracy` that won't be confused with the per-epoch logs.

another approach is to log the model separately after the final metrics are logged, just to ensure the association is clear:

```python
# after training and logging final metrics
mlflow.pytorch.log_model(model, name=model_type)
```

if you want to dig deeper into why this happens, check out the mlflow documentation on logging: [mlflow logging](https://mlflow.org/docs/latest/tracking.html#logging-data-to-runs). it's always a good idea to double-check how the metrics are being aggregated and displayed.

curious to see if this helps clarify things for you! let me know how it goes.

**Comment by AbdoDameen** [comment]
The metrics shown next to the model in the registry are whatever was logged most recently to the run at the moment log_model() was called.

If you see first-epoch values next to the model, it means the run's metric state had those as the most recent values when log_model() ran. This can happen if your per-epoch logging uses the same metric keys but the order of operations means the first epoch's values are the last ones written before log_model.

**Fix:** Log the final metrics right before saving:

```python
# After your training loop
mlflow.log_metric("val_loss", final_loss)  
mlflow.log_metric("val_accuracy", final_acc)
mlflow.pytorch.log_model(model, artifact_path="model")
```

**Debug:** Check the Metrics tab (not the model card) of your run. The model card shows one snapshot. The Metrics tab shows every value logged over time - you will see all epochs there.
